# Chatbot Intent Identification - Rule-Based Classifier

This notebook implements a simple rule-based classifier for identifying intents in chatbot inputs.

## Step 1: Define Intents

Define a small set of intents: greeting, booking, complaint.

In [5]:
# Define intents
intents = ['greeting', 'booking', 'complaint']

## Step 2: Create Rules

Create simple 'if-then' rules based on keywords.

In [6]:
# Define rules
def classify_intent(text):
    text = text.lower()
    if any(word in text for word in ['hello', 'hi', 'hey']):
        return 'greeting'
    elif any(word in text for word in ['book', 'reserve', 'schedule']):
        return 'booking'
    elif any(word in text for word in ['problem', 'issue', 'complaint']):
        return 'complaint'
    else:
        return 'unknown'

## Step 3: Prepare Test Data

Create a small test dataset with labeled examples.

In [7]:
# Test data
test_data = [
    ("Hello there", "greeting"),
    ("Hi, how are you?", "greeting"),
    ("I want to book a table", "booking"),
    ("Can I reserve a seat?", "booking"),
    ("There's a problem with my order", "complaint"),
    ("I have an issue", "complaint"),
    ("What's the weather?", "unknown")
]

## Step 4: Test and Evaluate

Run the classifier on test data and calculate accuracy.

In [8]:
# Test the classifier
correct = 0
total = len(test_data)
for text, true_intent in test_data:
    predicted = classify_intent(text)
    print(f"Text: '{text}' -> Predicted: {predicted}, Actual: {true_intent}")
    if predicted == true_intent:
        correct += 1

accuracy = (correct / total) * 100
print(f"\nAccuracy: {accuracy:.2f}%")

Text: 'Hello there' -> Predicted: greeting, Actual: greeting
Text: 'Hi, how are you?' -> Predicted: greeting, Actual: greeting
Text: 'I want to book a table' -> Predicted: booking, Actual: booking
Text: 'Can I reserve a seat?' -> Predicted: booking, Actual: booking
Text: 'There's a problem with my order' -> Predicted: complaint, Actual: complaint
Text: 'I have an issue' -> Predicted: complaint, Actual: complaint
Text: 'What's the weather?' -> Predicted: unknown, Actual: unknown

Accuracy: 100.00%


In [10]:
# Calculate rule coverage (percentage of inputs not classified as 'unknown')
covered = sum(1 for text, _ in test_data if classify_intent(text) != 'unknown')
coverage = (covered / total) * 100
print(f"Rule Coverage: {coverage:.2f}%")

Rule Coverage: 85.71%


## Results

The classifier achieved an accuracy of 100.00% on the test set. Rule coverage was 85.71%, meaning 6 out of 7 inputs were handled by the rules, with 'unknown' correctly assigned to unmatched inputs.